### PR TITLE
AWS Glue worker_type feature

### LINKED ISSUE #8257: Feature request: "worker type" and "number of workers" arguments on aws_glue_job resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS Glue team has introduced new worker types to support memory intensive and ML workloads on Glue. This brings the requirement to support "worker type" and "number of workers" agrumnets on the aws_glue_job resource.

See the announcement: https://aws.amazon.com/about-aws/whats-new/2019/04/aws-glue-now-supports-additional-configuration-options-for-memory-intensive-jobs/

### New or Affected Resource(s)

* aws_glue_job

### Potential Terraform Configuration

```hcl
resource "aws_glue_job" "job_id" {
  name                        = "job-name"
  worker_type             = "G.1X"
  number_of_workers = 10
  # other arguments
}
```

### References

* N/A

### PR DESCRIPTION
**Pull request by Zambonilli**
I have little experience in Google Go and would like an experienced set of eyes here. Especially around the logic for `worker_type` & `number_of_workers` being required/optional in opposite of `max_allocation` or `allocated_capacity` fields.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #8257 

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
Supports AWS Glue optional worker type by adding the `worker_type` and `number_of_workers` parameters. `worker_type` parameter can be set to `Standard`, `G.1X` or `G.2x`.
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSGlueJob_WorkerType'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSGlueJob_WorkerType -timeout 120m
=== RUN   TestAccAWSGlueJob_WorkerType
=== PAUSE TestAccAWSGlueJob_WorkerType
=== CONT  TestAccAWSGlueJob_WorkerType
--- PASS: TestAccAWSGlueJob_WorkerType (39.10s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	39.132s

$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSGlueJob'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSGlueJob -timeout 120m
=== RUN   TestAccAWSGlueJob_Basic
=== PAUSE TestAccAWSGlueJob_Basic
=== RUN   TestAccAWSGlueJob_AllocatedCapacity
=== PAUSE TestAccAWSGlueJob_AllocatedCapacity
=== RUN   TestAccAWSGlueJob_Command
=== PAUSE TestAccAWSGlueJob_Command
=== RUN   TestAccAWSGlueJob_DefaultArguments
=== PAUSE TestAccAWSGlueJob_DefaultArguments
=== RUN   TestAccAWSGlueJob_Description
=== PAUSE TestAccAWSGlueJob_Description
=== RUN   TestAccAWSGlueJob_ExecutionProperty
=== PAUSE TestAccAWSGlueJob_ExecutionProperty
=== RUN   TestAccAWSGlueJob_MaxRetries
=== PAUSE TestAccAWSGlueJob_MaxRetries
=== RUN   TestAccAWSGlueJob_Timeout
=== PAUSE TestAccAWSGlueJob_Timeout
=== RUN   TestAccAWSGlueJob_SecurityConfiguration
=== PAUSE TestAccAWSGlueJob_SecurityConfiguration
=== RUN   TestAccAWSGlueJob_WorkerType
=== PAUSE TestAccAWSGlueJob_WorkerType
=== RUN   TestAccAWSGlueJob_PythonShell
=== PAUSE TestAccAWSGlueJob_PythonShell
=== RUN   TestAccAWSGlueJob_MaxCapacity
=== PAUSE TestAccAWSGlueJob_MaxCapacity
=== CONT  TestAccAWSGlueJob_Basic
=== CONT  TestAccAWSGlueJob_WorkerType
=== CONT  TestAccAWSGlueJob_PythonShell
=== CONT  TestAccAWSGlueJob_ExecutionProperty
=== CONT  TestAccAWSGlueJob_SecurityConfiguration
=== CONT  TestAccAWSGlueJob_Timeout
=== CONT  TestAccAWSGlueJob_MaxCapacity
=== CONT  TestAccAWSGlueJob_MaxRetries
=== CONT  TestAccAWSGlueJob_DefaultArguments
=== CONT  TestAccAWSGlueJob_Description
=== CONT  TestAccAWSGlueJob_Command
=== CONT  TestAccAWSGlueJob_AllocatedCapacity
--- PASS: TestAccAWSGlueJob_Basic (30.01s)
--- PASS: TestAccAWSGlueJob_ExecutionProperty (35.53s)
--- PASS: TestAccAWSGlueJob_MaxRetries (41.82s)
--- PASS: TestAccAWSGlueJob_PythonShell (41.87s)
--- PASS: TestAccAWSGlueJob_MaxCapacity (52.79s)
--- PASS: TestAccAWSGlueJob_Timeout (63.17s)
--- PASS: TestAccAWSGlueJob_DefaultArguments (69.60s)
--- PASS: TestAccAWSGlueJob_Command (77.57s)
--- PASS: TestAccAWSGlueJob_WorkerType (83.86s)
--- PASS: TestAccAWSGlueJob_Description (85.55s)
--- PASS: TestAccAWSGlueJob_SecurityConfiguration (86.51s)
--- PASS: TestAccAWSGlueJob_AllocatedCapacity (133.35s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	133.406s
...
```

### CODE REVIEW

**Comment by avicennax** [comment]
@Zambonilli, do you know what the hold up is here on getting this merged in?

**Comment by monnetchr** [comment]
@Zambonilli maybe remove [WIP] from issue title?

**Comment by Zambonilli** [comment]
There's 325 open PRs currently for this project so I could only imagine how hard it is to manage that many PRs. Let's cut the team some slack here. I have WIP in the title because I'd like a 2nd set of eyes on this with more experience in Google Go. Can you be that 2nd set of eyes @monnetchr or @avicennax?

**Comment by avicennax** [comment]
Sure thing, but mind, I'm not terribly experienced in Go either. 
Not pointing fingers here btw haha. I just want to understand why after two months, your seemingly correct and complete work hasn't had so much as a follow up note. I think @monnetchr is on to something, lets remove [WIP] from the title and ping the gatekeeper(s).

**Comment by Zambonilli** [comment]
Updated title to remove [WIP] tag.

**Comment by stephencsmall** [comment]
Anything need to be done to get these checks to pass? Looks like one was cancelled.

**Comment by aeschright** [comment]
It's ok to ignore the pr-filter-sync task. As noted above, we're catching up on a pretty hefty PR backlog so I can give this a quick glance but it may take a bit longer to give it a full review. Thanks for your patience!

File: website/docs/r/glue_job.html.markdown [hunk 1]
```diff
@@ -60,7 +60,9 @@ be removed in future releases, please use `max_capacity` instead.
 * `name` – (Required) The name you assign to this job. It must be unique in your account.
 * `role_arn` – (Required) The ARN of the IAM role associated with this job.
 * `timeout` – (Optional) The job timeout in minutes. The default is 2880 minutes (48 hours).
-* `security_configuration` - (Optional) The name of the Security Configuration to be associated with the job. 
+* `security_configuration` - (Optional) The name of the Security Configuration to be associated with the job.
+* `worker_type` - (Optional) The type of predefined worker that is allocated when a job runs. Accepts a value of Standard, G.1X, or G.2X.
+* `number_of_workers` - (Optional) he number of workers of a defined workerType that are allocated when a job runs.
```
Comment (aeschright) [inline website/docs/r/glue_job.html.markdown]:
```suggestion
* `number_of_workers` - (Optional) The number of workers of a defined workerType that are allocated when a job runs.
```

**Comment by Zambonilli** [comment]
updated the documentation per code review via @aeschright

**Comment by Zambonilli** [comment]
This PR should be gtg.

**Comment by amitsehgal** [comment]
any plans of merging this PR ?

**Comment by Zambonilli** [comment]
ping on this getting merged

**Comment by kabeersvohra** [comment]
Can this be merged?

**Comment by Zambonilli** [comment]
yes

File: aws/resource_aws_glue_job.go [hunk 2]
```diff
@@ -113,6 +114,18 @@ func resourceAwsGlueJob() *schema.Resource {
 				Type:     schema.TypeString,
 				Optional: true,
 			},
+			"worker_type": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"allocated_capacity", "max_capacity", "allocated_capacity"},
+				ValidateFunc:  validation.StringMatch(regexp.MustCompile(`^Standard|G.1X|G.2X`), "Must be one of Standard, G.1X or G.2X. See https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-jobs-job.html"),
```
Comment (bflad) [inline aws/resource_aws_glue_job.go]: Nit: We typically prefer to use the [`validation.StringInSlice()`](https://godoc.org/github.com/hashicorp/terraform-plugin-sdk/helper/validation#StringInSlice) functionality coupled with available [AWS Go SDK constants](https://docs.aws.amazon.com/sdk-for-go/api/service/glue/#pkg-constants) for this type of validation, to limit the usage of regular expressions while also setting us up for future auto-generation opportunities, e.g.

```suggestion
				ValidateFunc:  validation.StringInSlice([]string{
					glue.WorkerTypeG1x,
					glue.WorkerTypeG2x,
					glue.WorkerTypeStandard,
				}, false),
```

File: aws/resource_aws_glue_job.go [hunk 3]
```diff
@@ -113,6 +114,18 @@ func resourceAwsGlueJob() *schema.Resource {
 				Type:     schema.TypeString,
 				Optional: true,
 			},
+			"worker_type": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"allocated_capacity", "max_capacity", "allocated_capacity"},
```
Comment (bflad) [inline aws/resource_aws_glue_job.go]: Nit: `allocated_capacity` is declared twice

File: aws/resource_aws_glue_job.go [hunk 4]
```diff
@@ -113,6 +114,18 @@ func resourceAwsGlueJob() *schema.Resource {
...
+				ValidateFunc:  validation.StringMatch(regexp.MustCompile(`^Standard|G.1X|G.2X`), "Must be one of Standard, G.1X or G.2X. See https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-jobs-job.html"),
+			},
+			"number_of_workers": {
+				Type:          schema.TypeInt,
+				Optional:      true,
+				ConflictsWith: []string{"allocated_capacity", "max_capacity", "allocated_capacity"},
```
Comment (bflad) [inline aws/resource_aws_glue_job.go]: Nit: `allocated_capacity` is declared twice

File: aws/resource_aws_glue_job.go (hunk 2 above)
Comment (jValdron) [reply aws/resource_aws_glue_job.go]: This is not formatted through `make fmt`. There is two spaces vetween `ValidateFunc:` and `validation`. It's currently making all PR tests fail.

Comment (Zambonilli) [reply aws/resource_aws_glue_job.go]: I think I ran `go fmt` and not `make fmt` from the root per the CONTRIBUTING.md's Enhancement/Bugfix to a Resource -> Well-formed Code section. I can run `make fmt` and push to the PR.

Comment (jValdron) [reply aws/resource_aws_glue_job.go]: No problem, fixed by 4e449b0

Comment (Zambonilli) [reply aws/resource_aws_glue_job.go]: thank you, sorry for an inconvenience!

**Pull request merged**